sympy/simplify/cse_main.py

```diff
@@ -81,11 +270,11 @@ def cse_separate(r, e):
     >>> from sympy import cos, exp, cse, Eq, symbols
     >>> x0, x1 = symbols('x:2')
     >>> eq = (x + 1 + exp((x + 1)/(y + 1)) + cos(y + 1))
-    >>> cse([eq, Eq(x, z + 1), z - 2], postprocess=cse_separate) in [
```

Comment:
Why is this not considered a backwards incompatibility? This could break someone's current scripts, especially considering it's in an example of how `cse` is supposed to be used.

Reply:
Thanks for bringing this to my attention. As `cse` is the only function that is part of the main SymPy namespace, I (perhaps wrongly) assumed that this function and other CSE functions in this module were "private". There is an inconsistency in returns types between these functions. Some (like `cse`, `tree_cse`) return `Tuple[List[Tuple[Symbol, Expr]], List[Expr]]`, others (like `cse_separate`) return `List[List[Tuple[Symbol, Expr]], List[Expr]]`. This doctest example is interesting as, prior to this PR, it does cause `cse` to return a `List[...]` instead of a `Tuple[...]`.

I'd argue that `cse_separate` returning a different type to `cse` and `tree_cse` is a bug. I believe that all CSE-related functions should return the same type, which is what this PR aims to address. But this comment has highlighted that making this change would indeed be backwards incompatible for instances where `cse_separate` is expected to specifically return a list.

I suggest changing `cse_separate` and `tree_cse` (and potentially `_cse_homogeneous` too, although this is "private" so matters less) to return a `CseExpr`.

Reply:
Last time I've seen the issue of "private" vs "public" functions came up, I think the conclusion was that there is no such formal distinction in SymPy, so unless something is explicitly stated to be private (e.g. in the docstring, or at the top of the module), then one can not assume a user could have been aware that it is not a public function. It could be that this policy has changed, since then.

Regardless, considering it a bug does not solve the problem that user's scripts suddenly stop working just because it was a more convenient solution to SymPy developers.

Perhaps a version of `CseExpr` can be made which inherits from `list` rather than `tuple`? E.g. `CseExprTuple` and `CseExprList`.